File: packages/membership-sdk/src/types/subscription.ts
import type { PublicKey } from "@solana/web3.js";

/**
 * On-chain Subscription structure
 */
export type SubscriptionAccount = {
  /** Subscription ID (same as PDA public key) */
  subscriptionId: PublicKey;
  /** Plan this subscription belongs to */
  plan: PublicKey;
  /** Encrypted user commitment */
  encryptedUserCommitment: Uint8Array;
  /** Membership commitment hash */
  membershipCommitment: Uint8Array;
  /** Unix timestamp when subscribed */
  subscribedAt: bigint;
  /** Unix timestamp when cancelled (0 if active) */
  cancelledAt: bigint;
  /** Whether the subscription is active */
  isActive: boolean;
  /** Encryption nonce */
  nonce: bigint;
  /** PDA bump seed */
  bump: number;
};

/**
 * Input for subscribing to a plan
 */
export type SubscribeInput = {
  /** Plan to subscribe to */
  plan: PublicKey;
  /** User's secret for commitment generation */
  userSecret: Uint8Array;
};

/**
 * Subscription with additional metadata
 */
export type Subscription = {
  /** Public key of the subscription account PDA */
  publicKey: PublicKey;
  /** Plan this subscription belongs to */
  plan: PublicKey;
  /** Membership commitment (for verification) */
  membershipCommitment: Uint8Array;
  /** Unix timestamp when subscribed */
  subscribedAt: Date;
  /** Unix timestamp when cancelled (null if active) */
  cancelledAt: Date | null;
  /** Whether the subscription is active */
  isActive: boolean;
};

/**
 * Subscription list filter options
 */
export type SubscriptionFilter = {
  /** Filter by plan */
  plan?: PublicKey;
  /** Filter by active status */
  isActive?: boolean;
  /** Maximum number of results */
  limit?: number;
  /** Offset for pagination */
  offset?: number;
};

/**
 * Membership proof for verification
 */
export type MembershipProof = {
  /** The subscription public key */
  subscription: PublicKey;
  /** The plan public key */
  plan: PublicKey;
  /** Membership commitment */
  commitment: Uint8Array;
  /** User's secret (never shared on-chain) */
  secret: Uint8Array;
};

File: subly_devnet/programs/subly_devnet/src/state/subscription.rs
use anchor_lang::prelude::*;

/// Subscription - stores subscription contract information
/// PDA seeds: ["subscription", plan, user_commitment]
#[account]
pub struct Subscription {
    /// Subscription's own address (for reference)
    pub subscription_id: Pubkey,
    /// Reference to plan account
    pub plan: Pubkey,
    /// Encrypted user commitment (Arcium encrypted)
    /// user_commitment = hash(secret || plan_id)
    pub encrypted_user_commitment: [u8; 32],
    /// Membership commitment for Light Protocol ZK proofs
    pub membership_commitment: [u8; 32],
    /// Subscription creation timestamp (plaintext - for backward compatibility, will be deprecated)
    pub subscribed_at: i64,
    /// Cancellation timestamp (0 if not cancelled) (plaintext - for backward compatibility, will be deprecated)
    pub cancelled_at: i64,
    /// Whether the subscription is active (plaintext - for backward compatibility, will be deprecated)
    pub is_active: bool,
    /// Nonce for encryption
    pub nonce: u128,
    /// PDA bump
    pub bump: u8,
    /// Encrypted subscription status (Arcium MXE encrypted)
    /// Contains: is_active (u8), subscribed_at (i64), cancelled_at (i64)
    pub encrypted_status: [u8; 64],
    /// Nonce for status encryption
    pub status_nonce: [u8; 16],
}

impl Subscription {
    /// Calculate account space
    pub const SPACE: usize = 8    // discriminator
        + 32                       // subscription_id: Pubkey
        + 32                       // plan: Pubkey
        + 32                       // encrypted_user_commitment: [u8; 32]
        + 32                       // membership_commitment: [u8; 32]
        + 8                        // subscribed_at: i64
        + 8                        // cancelled_at: i64
        + 1                        // is_active: bool
        + 16                       // nonce: u128
        + 1                        // bump: u8
        + 64                       // encrypted_status: [u8; 64]
        + 16;                      // status_nonce: [u8; 16]
    // Total: 250 bytes
}
